### ISSUE TITLE
Global path exclude

### ISSUE THREAD

**Issue by sanmai-NL** [evilmartians/lefthook#593]
### :zap: Summary

In monorepos, sometimes the CI/CD jobs running Lefthook must be specific to particular products, each included in that repo. Currently, Lefthook has no convenient way to globally include or exclude files based on criteria other than what's staged.

### Value

Also, when CI/CD jobs run on the default branch, for example, Lefthook may be invoked with `--all-files`. Jobs specific to a product and running on non-default branches should only include files related to that product, based on which files were staged/committed.

### Behavior and configuration changes

Support `--exclude` CLI parameter(s) that filter all file lists that Lefthook uses, even with `--all-files`, `{files}` or `{staged_files}`.

### Design and alternatives

One feature, perhaps larger epic, is to support parsing `CODEOWNERS`. It's possible to create sections in there that cover one or more globs. These sections could constitute products in a monorepo. Moreover, this adds a lot of value to give specific feedback to individual developers based on their role or product responsibility, when using Lefthook as local Git hook.

Alternatively, irrelevant files/directories could be deleted or be made inaccessible from the filesystem in the CI job, before launching Lefthook. However, that leads to unstaged changes that may influence some common commands (e.g., a command to forbid unstaged changes pre-push). Sometimes files are shared between products, so they can't be removed surgically. Also, some tools, like Prettier report failure when provided with globs to paths it can't read.

Also alternatively, the Lefthook configuration file(s) could be mutated in the CI job prior to Lefthook execution, so that files are excluded somehow. That is hard to maintain and sometimes hard to implement, when using a mixture of `{files}`, `{staged_files}` and different commands etc.

**Comment by scop** [comment]
Thumbs up for this. Especially important for my use case would be global `exclude:` support in configs.

**Comment by mrexox** [comment]
Hey! Do I correctly understand the feature request?
1. A global `exclude:` option. A hook-level `exclude` option which can be **extended** by command-specific `exclude` options?
```yml
pre-commit:
  exclude:
    - info/*.txt
    - docs/*.md
  commands:
    some-command:
      run: fix-typos {staged_files}
      exclude:
       - api/*.md # appended to `info/*.txt` and `docs/*.md` list of files to exclude
```

2. An `--exclude` option which will support a list of files to exclude. Does not overwrite, but **extends** the `exclude` options in the config.

```bash
# Also excludes `README.md` and `CHANGELOG.md` from any of `{files}`, `{staged_files}`, or `{push_files}`
lefthook run pre-commit --exclude README.md --exclude CHANGELOG.md
```

**Comment by sanmai-NL** [comment]
Yes!

**Comment by mrexox** [comment]
Sorry for the long delay. I've added the arg `--exclude` in 1.11.15 release

**Comment by scop** [comment]
Thanks for working on this! However it seems to me that only the `--exclude` option was added, and not the global (or hook scoped) `exclude:` configuration option? At least https://lefthook.dev/configuration/ does not document it.

If so, it would be nice to reopen this so that the config option won't be forgotten. For me, that's a more interesting feature than the command line option, because it automatically causes right things to happen, no matter if run from a hook, manually, or from a CI pipeline without having to duplicate the cmdline option to all (and in case of manual, having to actually remember to type it).

**Comment by mrexox** [comment]
Fair enough! I forgot about the global exclude option. @scop, would a hook-level `exclude` option be enough? Like in my example in a [comment](https://github.com/evilmartians/lefthook/issues/593#issuecomment-2782805510)

**Comment by scop** [comment]
That would be a nice improvement, thanks for considering! For my use cases global vs hook level doesn't really matter that I can remember. I just thought a global one could be easier to implement now that the global --exclude cmdline flag exists, but maybe it makes no difference.
